File: python/quarto/gameobject/piece.py
import numpy as np

class Piece:
    def __init__(self, color, shape, top, height):
        #色,形,穴,高さ
        self.param = np.array([color,shape,top,height])
        
        #文字列で格納された値を0,1に変換する
        c   = 1 if color  == 1      else 0
        s   = 1 if shape  == 1      else 0
        t   = 1 if top    == 1      else 0
        h   = 1 if height == 1      else 0

        #0,1からインデックスを作成する
        self.index = (c<<0) + (s<<1) + (t<<2) + (h<<3)
    
    def toDict(self):
        c = "light"     if self.param[0]  == 1 else "dark"
        s = "circular"  if self.param[1]  == 1 else "square"
        t = "hollow"    if self.param[2]  == 1 else "solid"
        h = "tall"      if self.param[3]  == 1 else "short"
        return {\
            'color':c,\
            'shape':s,\
            'top':t,\
            'height':h,\
        }

    def toNumList(self):
        return np.where(self.param==1,1,2)
    
    @classmethod
    def getInstance(cls, pieceDict):
        #文字列で格納された値を0,1に変換する
        c   = 1 if pieceDict['color']  == "light"     else 0
        s   = 1 if pieceDict['shape']  == "circular"  else 0
        t   = 1 if pieceDict['top']    == "hollow"    else 0
        h   = 1 if pieceDict['height'] == "tall"      else 0

        #0,1からインデックスを作成する
        index = (c<<0) + (s<<1) + (t<<2) + (h<<3)
        return __piece_list__[index]
    
    @classmethod
    def getAllPiece(cls):
        return list(__piece_list__)

__piece_list__ = (\
    Piece(-1,-1,-1,-1),\
    Piece( 1,-1,-1,-1),\
    Piece(-1, 1,-1,-1),\
    Piece( 1, 1,-1,-1),\
    Piece(-1,-1, 1,-1),\
    Piece( 1,-1, 1,-1),\
    Piece(-1, 1, 1,-1),\
    Piece( 1, 1, 1,-1),\
    Piece(-1,-1,-1, 1),\
    Piece( 1,-1,-1, 1),\
    Piece(-1, 1,-1, 1),\
    Piece( 1, 1,-1, 1),\
    Piece(-1,-1, 1, 1),\
    Piece( 1,-1, 1, 1),\
    Piece(-1, 1, 1, 1),\
    Piece( 1, 1, 1, 1),\
)

File: src/js/quarto/gameobject/piece.js
export class Piece{
    constructor(color, shape, top, height){
        //色,形,穴,高さ
        this.param = [color,shape,top,height];
        
        //文字列で格納された値を0,1に変換する
        let c   = (color  == 1)? 1:0;
        let s   = (shape  == 1)? 1:0;
        let t   = (top    == 1)? 1:0;
        let h   = (height == 1)? 1:0;

        //0,1からインデックスを作成する
        this.index = (c<<0) + (s<<1) + (t<<2) + (h<<3);
    }
    
    toDict(){
        let c = (this.param[0]  == 1)? "light"    : "dark";
        let s = (this.param[1]  == 1)? "circular" : "square";
        let t = (this.param[2]  == 1)? "hollow"   : "solid";
        let h = (this.param[3]  == 1)? "tall"     : "short";
        return {
            "color":c,
            "shape":s,
            "top":t,
            "height":h,
        };
    }

    toNumList(){
        //return np.where(this.param==1,1,2);
        let res = new Array(4);
        for(let i=0; i<4; ++i){
            res[i] = (this.param[i]==1)? 1:2;
        }
        return res;
    }
    
    static getInstance(pieceDict){
        //文字列で格納された値を0,1に変換する
        let c   = (pieceDict["color"]  == "light"   )? 1:0;
        let s   = (pieceDict["shape"]  == "circular")? 1:0;
        let t   = (pieceDict["top"]    == "hollow"  )? 1:0;
        let h   = (pieceDict["height"] == "tall"    )? 1:0;

        //0,1からインデックスを作成する
        let index = (c<<0) + (s<<1) + (t<<2) + (h<<3);
        return __piece_list__[index];
    }
    
    static getAllPiece(){
        return __piece_list__.slice();  //シャローコピー
    }
}

const __piece_list__ = [
    new Piece(-1,-1,-1,-1), //0
    new Piece( 1,-1,-1,-1), //1
    new Piece(-1, 1,-1,-1), //2
    new Piece( 1, 1,-1,-1), //3
    new Piece(-1,-1, 1,-1), //4
    new Piece( 1,-1, 1,-1), //5
    new Piece(-1, 1, 1,-1), //6
    new Piece( 1, 1, 1,-1), //7
    new Piece(-1,-1,-1, 1), //8
    new Piece( 1,-1,-1, 1), //9
    new Piece(-1, 1,-1, 1), //10
    new Piece( 1, 1,-1, 1), //11
    new Piece(-1,-1, 1, 1), //12
    new Piece( 1,-1, 1, 1), //13
    new Piece(-1, 1, 1, 1), //14
    new Piece( 1, 1, 1, 1), //15
];
